streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -83,10 +84,24 @@ public void register(final StateRestorer restorer) {
 
         final Set<TopicPartition> restoringPartitions = new HashSet<>(needsRestoring.keySet());
         try {
-            final ConsumerRecords<byte[], byte[]> allRecords = poll(restoreConsumer, 10);
-            for (final TopicPartition partition : restoringPartitions) {
-                restorePartition(allRecords, partition, active.restoringTaskFor(partition));
+            if (!needsRestoring.isEmpty()) {
```

Comment:
IMHO, we should add a second check to make sure `updatedEndOffsets` is set only once:
```
if (!needsRestoring.isEmpty() && updatedEndOffsets == null) {
```
Additionally, we should set `updatedEndOffsets = null` in Line 117 below (wondering if this is good enough, or if we need to reset `updatedEndOffsets = null` somewhere else, too. If we are in the middle of restore, and a rebalance happens, we get new partitions and need to refresh the corresponding endOffsets... \cc @guozhangwang WDYT?

Reply:
@mjsax What you suggested sounds right to me.

Reply:
Hi @mjsax . When I added the check as you suggested, I got the following:
org.apache.kafka.streams.processor.internals.StoreChangelogReaderTest > shouldRestorePartitionsRegisteredPostInitialization FAILED
    java.lang.NullPointerException
        at org.apache.kafka.streams.processor.internals.StoreChangelogReader.processNext(StoreChangelogReader.java:266)
        at org.apache.kafka.streams.processor.internals.StoreChangelogReader.restore(StoreChangelogReader.java:92)
        at org.apache.kafka.streams.processor.internals.StoreChangelogReaderTest.shouldRestorePartitionsRegisteredPostInitialization(StoreChangelogReaderTest.java:374).

Do you think you know the cause for this failure? I suggested something that might have been the cause in #4901, although that might not precisely be the source of the error.